hnn_core/dipole.py

```diff
@@ -152,62 +193,144 @@ class Dipole(object):
-        N_pyr_y : int
-            Number of Pyramidal cells in y direction
-        winsz : int
-            Smoothing window
+        window_len : int
+            Smoothing window in ms
         fctr : int
             Scaling factor
         """
-        self.baseline_renormalize(N_pyr_x, N_pyr_y)
-        self.convert_fAm_to_nAm()
         self.scale(fctr)
-        self.smooth(winsz)
 
-    def convert_fAm_to_nAm(self):
-        """ must be run after baseline_renormalization()
+        if window_len > 0:  # this is to allow param-files with len==0
+            self.smooth(window_len)
+
+    def _convert_fAm_to_nAm(self):
+        """The NEURON simulator output is in fAm, convert to nAm
+
+        NB! Must be run `after` :meth:`Dipole.baseline_renormalization`
         """
         for key in self.data.keys():
             self.data[key] *= 1e-6
-        self.units = 'nAm'
 
-    def scale(self, fctr):
+    def scale(self, factor):
+        """Scale (multiply) the dipole moment by a fixed factor
+
+        The attribute ``Dipole.scale_applied`` is updated to reflect factors
+        applied and displayed in plots.
+
+        Parameters
+        ----------
+        factor : int
+            Scaling factor, applied to the data in-place.
+        """
         for key in self.data.keys():
-            self.data[key] *= fctr
-        return fctr
-
-    def smooth(self, winsz):
-        # XXX: add check to make sure self.times is
-        # not smaller than winsz
-        if winsz <= 1:
-            return
+            self.data[key] *= factor
+        self.scale_applied *= factor
+        return self
+
+    def smooth(self, window_len):
+        """Smooth the dipole waveform using Hamming-windowed convolution
+
+        Note that this method operates in-place, i.e., it will alter the data.
+        If you prefer a filtered copy, consider using the
+        :meth:`~hnn_core.dipole.Dipole.copy`-method.
+
+        Parameters
+        ----------
+        window_len : float
+            The length (in ms) of a `~numpy.hamming` window to convolve the
+            data with.
+
+        Returns
+        -------
+        dpl_copy : instance of Dipole
+            A copy of the modified Dipole instance.
+        """
+        if not isinstance(window_len, (float, int)) or window_len < 0:
+            raise ValueError('Window length must be a non-negative number')
+        elif 0 < window_len < 1:
+            raise ValueError('Window length less than 1 ms is not supported')
+
+        # convolutional filter length is given in samples
+        winsz = np.round(1e-3 * window_len * self.sfreq)
+        if winsz > len(self.times):
+            raise ValueError(
+                f'Window length too long: {winsz} samples; data length is '
+                f'{len(self.times)} samples')
+
         for key in self.data.keys():
             self.data[key] = _hammfilt(self.data[key], winsz)
 
+        return self
+
+    def savgol_filter(self, h_freq):
```

Comment:
no test for this function?

Reply:
Only implicitly (`test_viz.py:L36`). Eric implemented some actual tests on this in `mne`, but I decided against copying those, since (again) we don't have that much data to make FFTs of...